### PR TITLE
Update linux compilers and fix OOM issues

### LINKED ISSUE #408: Figure out why builds are running out of memory
I'm guessing it's linking.

### PR DESCRIPTION
**Pull request by devpow112**
This drops `clang` **9** and adds `gcc` **11** and `clang` **11** and **12**. Resolves #408.

### CODE REVIEW

**Comment by devpow112** [comment]
Won't work till #412 merges.

**Comment by devpow112** [comment]
Looks like latest compilers have be warnings. Will see if I can address them

File: Overlays/Posix/Overlays/GCC/cmake/build_options.cmake [hunk 1]
```diff
@@ -7,6 +7,7 @@ add_compile_options(-Wno-stringop-truncation)
 add_compile_options(-Wno-unused-function)
 add_compile_options(-Wno-unused-variable)
 add_compile_options(-Wno-strict-aliasing)
+add_compile_options(-Wno-nonnull)
```
Comment (devpow112) [inline Overlays/Posix/Overlays/GCC/cmake/build_options.cmake:10]: New GCC has some issues with the script AST parser files but only in release

**Comment by devpow112** [comment]
GCC debug just uses so much memory when trying to list script test runner

File: Overlays/Posix/Overlays/GCC/cmake/build_options.cmake [hunk 2]
```diff
@@ -7,8 +7,13 @@ add_compile_options(-Wno-stringop-truncation)
...
 add_compile_options(-Wno-nonnull-compare)
 
+if(WN_LOW_RESOURCE_MODE)
```
Comment (devpow112) [inline Overlays/Posix/Overlays/GCC/cmake/build_options.cmake:13]: Hopefully fix linking issues

File: Overlays/Posix/Overlays/GCC/cmake/build_options.cmake [hunk 3]
```diff
@@ -7,8 +7,13 @@ add_compile_options(-Wno-stringop-truncation)
...
+  add_link_options(--reduce-memory-overheads)
+endif()
```
Comment (devpow112) [inline Overlays/Posix/Overlays/GCC/cmake/build_options.cmake:15]: Hopefully fix linking issues

Comment (devpow112) [reply Overlays/Posix/Overlays/GCC/cmake/build_options.cmake:15]: May also add `--no-keep-memory`

File: Overlays/Posix/Overlays/GCC/cmake/build_options.cmake [hunk 4]
```diff
@@ -7,8 +7,14 @@ add_compile_options(-Wno-stringop-truncation)
 add_compile_options(-Wno-unused-function)
 add_compile_options(-Wno-unused-variable)
 add_compile_options(-Wno-strict-aliasing)
+add_compile_options(-Wno-nonnull)
 add_compile_options(-Wno-nonnull-compare)
 
+if(WN_LOW_RESOURCE_MODE)
+  add_link_options($<$<CONFIG:Debug>:-Wl,--reduce-memory-overheads>)
+  add_link_options($<$<CONFIG:Debug>:-Wl,--no-keep-memory>)
+endif()
```
Comment (devpow112) [inline Overlays/Posix/Overlays/GCC/cmake/build_options.cmake:16]: This seems to have handled the OOM issues.

**Comment by devpow112** [comment]
I will try and re-enable SCCACHE as well but in another PR. This may have been the issue all along.

**Comment by devpow112** [comment]
Yay! 🎉 

Its-Working-Star-Wars

**Review by AWoloszyn — APPROVED** [review]

**Pull request merged**